### PR TITLE
Edit regex

### PR DESCRIPTION
**Pull request by andrzejgorski**

### CODE REVIEW

File: whylog_vim/output_formater/output_aggregator.py [hunk 1]
```diff
@@ -16,16 +17,23 @@ def get_content(self):
...
+    def call_by_menu_function(self, menu_function):
+        self.call_button(self.function_lines[menu_function])
+
+    def get_line_number(self, function_meta):
+        return self.function_lines[function_meta]
+
     def call_button(self, line_number):
         try:
-            self.buttons[line_number]()
+            self.callbacks[line_number]()
```
Comment (kgromadzki) [inline whylog_vim/output_formater/output_aggregator.py]: Refactor using get function

File: whylog_vim/output_formater/teacher_formater/__init__.py [hunk 2]
```diff
@@ -24,64 +24,72 @@ def format_parser(self, output, parser, parser_id, effect=True):
...
         output.add(parser.pattern)
-        output.create_button(partial(self.teacher_proxy.edit_regex, parser._id))
+        output.create_button(
+            partial(self.teacher_proxy.edit_regex, parser), (
+                FunctionNames.EDIT_REGEX, parser.line_id
+            )
+        )
         output.add(ParserOutputs.GUESS_BUTTON)
-        output.create_button(partial(self.teacher_proxy.guess_regex, parser._id))
+        # output.create_button(partial(self.teacher_proxy.guess_regex, parser))
```
Comment (kgromadzki) [inline whylog_vim/output_formater/teacher_formater/__init__.py]: Explain why this code is commented

Comment (andrzejgorski) [inline whylog_vim/output_formater/teacher_formater/__init__.py]: In this PR I change lot of interface. If Paweł accept it I will remove this comment. Now this code is not compatible so I commented it and wait for clues from Paweł.

File: whylog_vim/output_formater/output_aggregator.py [hunk 3]
```diff
@@ -16,21 +21,26 @@ def get_content(self):
...
+
     def call_button(self, line_number):
-        try:
-            self.buttons[line_number]()
-        except KeyError:
-            try:
-                self.default_callback_function()
-            except AttributeError:
-                pass
+        function = self.callbacks.get(line_number)
```
Comment (ppolewicz) [inline whylog_vim/output_formater/output_aggregator.py]:
``` python
callback = self.callbacks.get(line_number, self.default_callback)
return callback()
```

File: whylog_vim/output_formater/teacher_formater/__init__.py [hunk 4]
```diff
@@ -24,64 +24,72 @@ def format_parser(self, output, parser, parser_id, effect=True):
...
+                FunctionNames.EDIT_LINE_CONTENT, parser.line_id
+            )
         )
-        output.create_button(partial(self.teacher_proxy.edit_line_content, parser._id))
         output.add(ParserOutputs.META % (parser.line_resource_location, parser.line_offset))
         if not effect:
             output.add(ParserOutputs.COPY_BUTTON)
             output.add(ParserOutputs.DELETE_BUTTON)
-            output.create_button(partial(self.teacher_proxy.delete_parser, parser._id))
+            # output.create_button(partial(self.teacher_proxy.delete_parser, parser._id))
```
Comment (ppolewicz) [inline whylog_vim/output_formater/teacher_formater/__init__.py]: commented code

File: whylog_vim/output_formater/output_aggregator.py [hunk 5]
```diff
@@ -16,21 +21,26 @@ def get_content(self):
...
+    def create_button(self, callback_function, function_meta, line=None):
         line = line or self._get_current_line_number()
-        self.buttons[line] = callback_function
+        self.function_lines[function_meta] = line
+        self.callbacks[line] = callback_function
 
     def set_default_callback_function(self, default_callback_function):
         self.default_callback_function = default_callback_function
 
+    def call_by_menu_function(self, menu_function):
```
Comment (ppolewicz) [inline whylog_vim/output_formater/output_aggregator.py]: how will this be used?

File: whylog_vim/output_formater/output_aggregator.py [hunk 6]
```diff
@@ -16,21 +21,26 @@ def get_content(self):
...
+        self.call_button(self.function_lines[menu_function])
+
+    def get_line_number(self, function_meta):
```
Comment (ppolewicz) [inline whylog_vim/output_formater/output_aggregator.py]: what is this for?

**Comment by ppolewicz** [comment]
Could you please describe what are you trying to do with the buttons?

File: whylog_vim/output_formater/output_aggregator.py (hunk 5 above)
Comment (andrzejgorski) [inline whylog_vim/output_formater/output_aggregator.py]: In the future we will add shortcuts for every operation of teacher's menu. If someone will wants to edit line content, he will press `el0c` (edit line 0 content) and the whylog proxy will call function this function with parametr (FunctionNames.EDIT_LINE_CONTENT, 0).

File: whylog_vim/output_formater/output_aggregator.py (hunk 6 above)
Comment (andrzejgorski) [inline whylog_vim/output_formater/output_aggregator.py]: This funciton is to easier testing and for eventualy addtion of shortcuts for jump to the specific position in TeacherMenu. `jl0c` - will move cursor to the line where is content of line 0 (effect).

File: whylog_vim/output_formater/output_aggregator.py [hunk 7]
```diff
@@ -3,8 +3,13 @@
 
 class OutputAggregator(object):
     def __init__(self):
-        self.buttons = {}
+        def pass_function():
```
Comment (ppolewicz) [inline whylog_vim/output_formater/output_aggregator.py]: please move it to module scope

File: whylog_vim/output_formater/output_aggregator.py [hunk 8]
```diff
@@ -16,21 +21,23 @@ def get_content(self):
         # The same type as vim buffor.
         return self.output_lines
 
-    def create_button(self, callback_function, line=None):
+    def create_button(self, callback_function, function_meta, line=None):
```
Comment (ppolewicz) [inline whylog_vim/output_formater/output_aggregator.py:46]: please document what `self.function_lines` and `function_meta` are

File: whylog_vim/output_formater/output_aggregator.py (hunk 5 above)
Comment (ppolewicz) [inline whylog_vim/output_formater/output_aggregator.py]: please document it

File: whylog_vim/output_formater/output_aggregator.py (hunk 6 above)
Comment (ppolewicz) [inline whylog_vim/output_formater/output_aggregator.py]: `function_lines` is public so you can access it directly when testing. Please remove this function for now. When (if...) keyboard shortcuts will be added in the future, we'll add it or do it differently.

File: whylog_vim/output_formater/teacher_formater/__init__.py (hunk 2 above)
Comment (ppolewicz) [inline whylog_vim/output_formater/teacher_formater/__init__.py]: the line is good

File: whylog_vim/output_formater/teacher_formater/__init__.py [hunk 9]
```diff
@@ -24,64 +24,72 @@ def format_parser(self, output, parser, parser_id, effect=True):
...
+        # )
 
     def _format_converters(self, output, groups, parser_id):
         for group in groups.keys():
             output.add(
                 ParserOutputs.GROUP_CONVERTER %
                 (group, groups[group].converter, groups[group].content)
             )
-            output.create_button(partial(self.teacher_proxy.edit_converter, parser_id, group))
+            # output.create_button(partial(self.teacher_proxy.edit_converter, parser_id, group))
```
Comment (ppolewicz) [inline whylog_vim/output_formater/teacher_formater/__init__.py]: either pass objects everywere, or ids everywhere. Objects seem better.

File: whylog_vim/output_formater/teacher_formater/input_windows_messages.py [hunk 10]
```diff
@@ -69,3 +69,12 @@ def get_edit_line_message(cls, old_content):
         output.add_commented(Messages.ENDING)
         output.add(old_content)
         return output
+
+    @classmethod
+    def get_edit_regex(cls, line_content, old_regex):
```
Comment (ppolewicz) [inline whylog_vim/output_formater/teacher_formater/input_windows_messages.py]: please rename the function appropriately

File: whylog_vim/proxy/teacher/utils.py [hunk 11]
```diff
@@ -1,28 +1,33 @@
...
-get_next_parser_id = partial(next, count(0))
-get_next_constraints_id = partial(next, count(0))
-
 
 class MenuHandler(object):
-    def edit_line_content(self, parser_id):
-        output = InputMessages.get_edit_line_message(self.rule.parsers[parser_id].line_content)
+    def edit_line_content(self, parser):
+        output = InputMessages.get_edit_line_message(parser.line_content)
+        effect = self.rule.effect_id == parser.line_id
```
Comment (ppolewicz) [inline whylog_vim/proxy/teacher/utils.py]: `is_effect`

**Pull request merged**